File: src/main/java/network/reflected/rfnetapi/commands/CommandRegistry.java
package network.reflected.rfnetapi.commands;

import io.papermc.paper.event.player.AsyncChatEvent;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.TextComponent;
import org.bukkit.Bukkit;
import org.bukkit.command.CommandSender;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerCommandPreprocessEvent;
import org.bukkit.event.server.ServerCommandEvent;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CommandRegistry implements Listener {
    Map<String, Command> commands = new HashMap<>();

    public void registerCommand(CommandRunnable command, String permission, int numberOfArgs, String name) {
        if (!commands.containsKey(name)) {
            if (!Bukkit.getCommandMap().getKnownCommands().containsKey(name)) {
                Bukkit.getCommandMap().register("reflected", new EmptyCommand(name));
            }
            commands.put(name, new Command(command, permission, numberOfArgs));
        } else {
            throw new ArrayStoreException("That command has already been registered!");
        }
    }

    public void registerCommands(CommandRunnable command, String permission, int numberOfArgs, String... names ) {
        for (String name : names) {
            registerCommand(command, permission, numberOfArgs, name);
        }
    }

    public void registerCommand(CommandRunnable command, int numberOfArgs, String name) {
        registerCommand(command, "", numberOfArgs, name);
    }

    public void registerCommands(CommandRunnable command, int numberOfArgs, String... names ) {
        for (String name : names) {
            registerCommand(command, numberOfArgs, name);
        }
    }

    private boolean executeCommand(CommandSender commandSender, String commandStr) {
        List<String> tokenized = tokenize(commandStr);
        Command command = commands.get(tokenized.get(0));
        if (command == null) {
            return false;
        }

        if (!command.permission.equals("") && !commandSender.hasPermission(command.permission)) {
            return false; // If they don't have permission, pretend the command doesn't exist.
        }

        if (tokenized.size() - 1 != command.argCount) {
            commandSender.sendMessage(Component.text("This command expects " + command.argCount + " arguments."));
            return true;
        }

        tokenized.remove(0);
        command.getRunnable().run(
                commandSender,
                CommandArg.parse(tokenized.toArray(new String[0]))
        );
        return true;
    }

    private static List<String> tokenize(String input) {
        List<String> tokens = new ArrayList<>();
        boolean inquotes = false;
        StringBuilder currentToken = new StringBuilder();
        for (int i = 0; i < input.length(); i++) {
            switch (input.charAt(i)) {
                case ' ':
                    if (inquotes || (i > 0 && input.charAt(i-1) == '\\')) {
                        currentToken.append(' ');
                    } else if (!currentToken.toString().equals("")) {
                        tokens.add(currentToken.toString());
                        currentToken = new StringBuilder();
                    }
                    break;
                case '"':
                    if (i > 0 && input.charAt(i-1) == '\\') {
                        currentToken.append('"');
                    } else {
                        inquotes = true;
                    }
                    break;
                case '\\':
                    if (i > 0 && input.charAt(i-1) == '\\') {
                        currentToken.append('\\');
                    }
                    break;
                default:
                    currentToken.append(input.charAt(i));
                    break;
            }
        }
        if (!currentToken.toString().equals("")) tokens.add(currentToken.toString());
        return tokens;
    }

    @EventHandler
    private void playerCommand(AsyncChatEvent event) {
        if (((TextComponent)event.message()).content().startsWith("?")) {
            if (!executeCommand(event.getPlayer(), ((TextComponent)event.message()).content().substring(1))) {
                event.getPlayer().sendMessage(Component.text("Command not found."));
            }
            event.setCancelled(true);
        }
    }

    @EventHandler
    private void playerCommand(PlayerCommandPreprocessEvent event) {
        if (event.getMessage().startsWith("/")) {
            event.setCancelled(executeCommand(event.getPlayer(), event.getMessage().substring(1)));
        }
    }

    @EventHandler
    private void serverCommand(ServerCommandEvent event) {
        if (event.getCommand().startsWith("/")) {
            event.setCancelled(executeCommand(event.getSender(), event.getCommand().substring(1)));
        } else {
            event.setCancelled(executeCommand(event.getSender(), event.getCommand()));
        }
    }

    @RequiredArgsConstructor
    private static class Command {
        @Getter final CommandRunnable runnable;
        @Getter final String permission;
        @Getter final int argCount;
    }

    private static class EmptyCommand extends org.bukkit.command.Command {
        protected EmptyCommand(@NotNull String name) {
            super(name);
        }

        @Override
        public boolean execute(@NotNull CommandSender sender, @NotNull String commandLabel, @NotNull String[] args) {
            return true; // Ignore that this command exists
        }
    }
}

File: src/main/java/com/reflectednetwork/rfnetapi/commands/CommandRegistry.kt
package com.reflectednetwork.rfnetapi.commands

import io.papermc.paper.event.player.AsyncChatEvent
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.TextComponent
import net.kyori.adventure.text.format.NamedTextColor
import com.reflectednetwork.rfnetapi.bugs.ExceptionDispensary
import org.bukkit.Bukkit
import org.bukkit.command.CommandSender
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerCommandPreprocessEvent
import org.bukkit.event.server.ServerCommandEvent

class CommandRegistry : Listener {
    var commands: MutableMap<String, Command> = HashMap()
    fun registerCommand(command: (executor: CommandSender, arguments: CommandArguments) -> Unit, permission: String, numberOfArgs: Int, name: String) {
        if (!commands.containsKey(name)) {
            if (!Bukkit.getCommandMap().knownCommands.containsKey(name)) {
                Bukkit.getCommandMap().register("reflected", EmptyCommand(name))
            }
            commands[name] = Command(command, permission, numberOfArgs)
        } else {
            throw ArrayStoreException("That command has already been registered!")
        }
    }

    fun registerCommands(command: (executor: CommandSender, arguments: CommandArguments) -> Unit, permission: String, numberOfArgs: Int, vararg names: String) {
        for (name in names) {
            registerCommand(command, permission, numberOfArgs, name)
        }
    }

    fun registerCommand(command: (executor: CommandSender, arguments: CommandArguments) -> Unit, numberOfArgs: Int, name: String) {
        registerCommand(command, "", numberOfArgs, name)
    }

    fun registerCommands(command: (executor: CommandSender, arguments: CommandArguments) -> Unit, numberOfArgs: Int, vararg names: String) {
        for (name in names) {
            registerCommand(command, numberOfArgs, name)
        }
    }

    private fun executeCommand(commandSender: CommandSender, commandStr: String): Boolean {
        val tokenized = tokenize(commandStr)
        val command = commands[tokenized[0]] ?: return false

        if (command.permission != "" && !commandSender.hasPermission(command.permission!!)) {
            commandSender.sendMessage(Component.text("That command is registered, but it doesn't do anything yet! Please use /help for working commands.").color(NamedTextColor.RED))
            return false // If they don't have permission, pretend the command doesn't exist.
        }

        if (tokenized.size - 1 != command.argCount) {
            commandSender.sendMessage(Component.text("This command expects " + command.argCount + " arguments.").color(NamedTextColor.RED))
            return true
        }

        tokenized.drop(0)
        try {
            command.runnable.invoke(
                commandSender,
                CommandArguments(tokenized)
            )
        } catch (error: ArgumentParseException) {
            commandSender.sendMessage(Component.text(error.errorText).color(NamedTextColor.RED))
        } catch (e: Exception) {
            ExceptionDispensary.reportAndNotify(e, "dispatching command", commandSender)
        }
        return true
    }

    @EventHandler
    private fun playerCommand(event: AsyncChatEvent) {
        if ((event.message() as TextComponent).content().startsWith("?")) {
            if (!executeCommand(event.player, (event.message() as TextComponent).content().substring(1))) {
                event.player.sendMessage(Component.text("Command not found."))
            }
            event.isCancelled = true
        }
    }

    @EventHandler
    private fun playerCommand(event: PlayerCommandPreprocessEvent) {
        if (event.message.startsWith("/")) {
            event.isCancelled = executeCommand(event.player, event.message.substring(1))
        }
    }

    @EventHandler
    private fun serverCommand(event: ServerCommandEvent) {
        if (event.command.startsWith("/")) {
            event.isCancelled = executeCommand(event.sender, event.command.substring(1))
        } else {
            event.isCancelled = executeCommand(event.sender, event.command)
        }
    }

    private class EmptyCommand(name: String) : org.bukkit.command.Command(name) {
        override fun execute(sender: CommandSender, commandLabel: String, args: Array<String>): Boolean {
            return true // Ignore that this command exists
        }
    }

    private fun tokenize(input: String): List<String> {
        val tokens: MutableList<String> = ArrayList()
        var inquotes = false
        var currentToken = StringBuilder()
        for (i in input.indices) {
            when (input[i]) {
                ' ' -> if (inquotes || i > 0 && input[i - 1] == '\\') {
                    currentToken.append(' ')
                } else if (currentToken.toString() != "") {
                    tokens.add(currentToken.toString())
                    currentToken = StringBuilder()
                }
                '"' -> if (i > 0 && input[i - 1] == '\\') {
                    currentToken.append('"')
                } else {
                    inquotes = true
                }
                '\\' -> if (i > 0 && input[i - 1] == '\\') {
                    currentToken.append('\\')
                }
                else -> currentToken.append(input[i])
            }
        }
        if (currentToken.toString() != "") tokens.add(currentToken.toString())
        return tokens
    }
}